erpnext/accounts/doctype/payment_manager/payment_manager.py

```diff
@@ -0,0 +1,429 @@
+			if (entry.payment_reference is None):
+				entry.payment_reference = entry.description
+			entry.invoices = ",".join(matching_invoices)
+			#print("Matching payment is {0}:{1}".format(entry.reference_type, entry.reference_name))
+
+	def get_matching_payments(self, party, amount, pay_date):
+		query = """select 'Payment Entry' as reference_type, name as reference_name, paid_amount as amount
+					from `tabPayment Entry` where party='{0}' and paid_amount={1} and posting_date='{2}' and docstatus != 2
+					""".format(party, amount, pay_date)
+		matching_payments = frappe.db.sql(query, as_dict=True)
+		return matching_payments
+
+	def map_unknown_transactions(self):
+		for entry in self.new_transaction_items:
+			if (entry.party): continue
+			inv_type = "Sales Invoice" if (entry.amount > 0) else "Purchase Invoice"
+			party_type = "customer" if (entry.amount > 0) else "supplier"
+
+			query = """select posting_date, name, {0}, outstanding_amount
+							from `tab{1}` where ROUND(outstanding_amount)={2} and posting_date < '{3}'
+							""".format(party_type, inv_type, round(abs(entry.amount)), entry.transaction_date)
+			invoices = frappe.db.sql(query, as_dict = True)
+			if(len(invoices) > 0):
+				entry.party = invoices[0].get(party_type)
+
+	def populate_matching_vouchers(self):
+		for entry in self.new_transaction_items:
+			if (not entry.party or entry.reference_name): continue
+			print("Finding matching voucher for {0}".format(entry.description))
+			amount = abs(entry.amount)
+			invoices = []
+			vouchers = get_matching_journal_entries(self.from_date, self.to_date, entry.party, self.bank_account, amount)
+			if len(vouchers) == 0: continue
+			for voucher in vouchers:
+				added = next((entry.invoice for entry in self.payment_invoice_items if entry.invoice == voucher.voucher_no), None)
+				if (added):
+					print("Found voucher {0}".format(added))
+					continue
+				print("Adding voucher {0} {1} {2}".format(voucher.voucher_no, voucher.posting_date, voucher.debit))
+				ent = self.append('payment_invoice_items', {})
+				ent.invoice_date = voucher.posting_date
+				ent.invoice_type = "Journal Entry"
+				ent.invoice = voucher.voucher_no
+				ent.payment_description = entry.description
+				ent.allocated_amount = max(voucher.debit, voucher.credit)
+
+				invoices += [ent.invoice_type + "|" + ent.invoice]
+				entry.reference_type = "Journal Entry"
+				entry.mode_of_payment = "Wire Transfer"
+				entry.reference_name = ent.invoice
+				#entry.account = entry.party
+				entry.invoices = ",".join(invoices)
+				break
+
+
+	def create_payment_entries(self):
+		for payment_entry in self.new_transaction_items:
+			if (not payment_entry.party): continue
+			if (payment_entry.reference_name): continue
+			print("Creating payment entry for {0}".format(payment_entry.description))
+			if (payment_entry.party_type == "Account"):
+				payment = self.create_journal_entry(payment_entry)
+				invoices = [payment.doctype + "|" + payment.name]
```

Comment:
Best to use JSON encoding to store such values. 

Not sure why we need this in the first place

Reply:
I don't recollect the exact reason, a comment would have helped here :( I think I was following the similar format in existing payment reconciliation tool. Let me see what I can improve.